mondrian/src/main/java/mondrian/rolap/agg/SegmentBuilder.java

```diff
@@ -881,6 +883,63 @@ public SegmentWithData convert(
             return addData(segment, body);
         }
     }
+    
+    /**
+     * Converts a segment's CellKey into axis values
+     * so that they can be compared across segments.    
+     * 
+     * Ex. (0, 2, 0) to [1997, 21, M]
+     */
+    private static class ColumnValues implements Comparable {
```

Comment:
I think this class needs to override equals() and hashCode()

Reply:
This class was created so that we could check for duplicate tuples between segments using a sorted compare operation instead of a slow tuple by tuple equals check.  You can see here how I switch from an ArrayList to a TreeSet:

https://github.com/pentaho/mondrian/pull/1221/files/4c0d5bfcc72b339600e179c758f4ce865ecbaeaf#diff-dfe16789b67477bd98be9ab8ae3886daR303

Using ArrayList to check for duplicates was the root cause of the bottleneck.

Reply:
Chatted with Luc and I'll add hashCode and equals and have them throw UnsupportedOperationException to indicate that this new class is not to be used in an ArrayList or Set.